File: typescript/tracing/respan-tracing-sdk/multi_processor.ts
import { RespanTelemetry } from '@respan/tracing';
import { updateCurrentSpan, addSpanEvent } from '@respan/tracing';
import * as fs from 'fs';
import * as path from 'path';
import dotenv from 'dotenv';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

dotenv.config({ path: path.join(__dirname, '.env') });

/**
 * This example demonstrates span tracking and custom logging:
 * 1. Tracking different types of tasks (normal, debug, analytics, slow)
 * 2. Adding custom attributes and events (via `updateCurrentSpan`, `addSpanEvent`)
 * 3. Demonstrating a multi-processor pipeline with:
 *    - routing (different processors handle different span types)
 *    - filters (processors decide which spans they accept)
 *    - multiple exporters (console + file exporters)
 *
 * Note: The public README references an `addProcessor` API, but not all SDK
 * versions expose a stable processor hook yet. To keep the example accurate and
 * runnable across versions, this file implements a small "processor pipeline"
 * in userland that consumes the same span metadata we attach via the SDK.
 */

type SpanExportRecord = {
  name: string;
  type: string;
  durationMs: number;
  startedAtIso: string;
  endedAtIso: string;
  attributes?: Record<string, unknown>;
  events?: Array<{
    name: string;
    attributes?: Record<string, unknown>;
    timestampMs: number;
  }>;
};

type SpanExporter = {
  name: string;
  export: (record: SpanExportRecord) => void | Promise<void>;
};

type SpanProcessor = {
  name: string;
  /** Return true when this processor should receive the span record. */
  matches: (record: SpanExportRecord) => boolean;
  exporters: SpanExporter[];
};

// Helper to log span information to file
function logSpanToFile(filepath: string, spanInfo: any) {
  try {
    const dir = path.dirname(filepath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
    fs.appendFileSync(filepath, JSON.stringify(spanInfo) + '\n');
    console.log(`  Logged to ${filepath}`);
  } catch (error) {
    console.error('  File logging error:', error);
  }
}

// Helper to log span information to console
function logSpanToConsole(prefix: string, spanInfo: any) {
  console.log(`  [${prefix}] ${spanInfo.name} - ${spanInfo.type}`);
}

function createFileExporter(filepath: string): SpanExporter {
  return {
    name: `file:${filepath}`,
    export: (record) => logSpanToFile(filepath, record),
  };
}

function createConsoleExporter(prefix: string): SpanExporter {
  return {
    name: `console:${prefix}`,
    export: (record) => logSpanToConsole(prefix, record),
  };
}

async function processWithProcessors(
  processors: SpanProcessor[],
  record: SpanExportRecord,
) {
  const matched = processors.filter((p) => {
    try {
      return p.matches(record);
    } catch {
      return false;
    }
  });

  if (matched.length === 0) {
    return;
  }

  // Fan-out: one record can be exported by multiple processors/exporters.
  await Promise.all(
    matched.flatMap((p) =>
      p.exporters.map(async (e) => {
        try {
          await e.export(record);
        } catch (error) {
          console.error(`  Exporter error (${p.name} -> ${e.name}):`, error);
        }
      }),
    ),
  );
}

async function runMultiProcessorDemo() {
  console.log('Starting Span Tracking & Logging Demo\n');

  const respan = new RespanTelemetry({
    apiKey: process.env.RESPAN_API_KEY || 'demo-key',
    baseURL: process.env.RESPAN_BASE_URL,
    appName: 'span-tracking-demo',
    logLevel: 'info',
    disableBatch: true,
  });

  await respan.initialize();
  console.log('Tracing initialized\n');

  // Processor setup: routing + filters + multiple exporters.
  const processors: SpanProcessor[] = [
    {
      name: 'debug-processor',
      matches: (r) => r.type === 'debug',
      exporters: [createFileExporter('./debug-spans.jsonl')],
    },
    {
      name: 'analytics-processor',
      matches: (r) => r.type === 'analytics',
      exporters: [
        createConsoleExporter('Analytics'),
        createFileExporter('./analytics-spans.jsonl'),
      ],
    },
    {
      name: 'slow-processor',
      matches: (r) => r.type === 'slow' || r.durationMs >= 150,
      exporters: [
        createConsoleExporter('SlowSpans'),
        createFileExporter('./slow-spans.jsonl'),
      ],
    },
    {
      name: 'default-console-processor',
      matches: (_r) => true,
      exporters: [createConsoleExporter('Default')],
    },
  ];

  await respan.withWorkflow({ name: 'span_tracking_workflow' }, async () => {
    // Normal task - standard tracing
    console.log('1. Normal Task:');
    await respan.withTask({ name: 'normal_task' }, async () => {
      const startTime = Date.now();
      updateCurrentSpan({
        attributes: {
          'task.type': 'normal',
          'task.priority': 'medium',
        },
      });
      addSpanEvent('task.started', { timestampMs: Date.now() });
      await new Promise((resolve) => setTimeout(resolve, 50));
      addSpanEvent('task.completed', { timestampMs: Date.now() });

      const now = Date.now();
      const record: SpanExportRecord = {
        name: 'normal_task',
        type: 'normal',
        durationMs: now - startTime,
        startedAtIso: new Date(startTime).toISOString(),
        endedAtIso: new Date(now).toISOString(),
        attributes: { 'task.type': 'normal', 'task.priority': 'medium' },
        events: [
          { name: 'task.started', timestampMs: startTime },
          { name: 'task.completed', timestampMs: now },
        ],
      };

      await processWithProcessors(processors, record);
      console.log('  Completed (processed via routing pipeline)');
    });

    // Debug task - with debug logging
    console.log('\n2. Debug Task:');
    await respan.withTask({ name: 'debug_task' }, async () => {
      const startTime = Date.now();
      updateCurrentSpan({
        attributes: {
          'task.type': 'debug',
          'task.priority': 'low',
          'debug.enabled': true,
        },
      });
      addSpanEvent('debug.started', { level: 'verbose' });
      
      await new Promise((resolve) => setTimeout(resolve, 50));
      
      const endTime = Date.now();
      const spanInfo = {
        name: 'debug_task',
        type: 'debug',
        durationMs: endTime - startTime,
        startedAtIso: new Date(startTime).toISOString(),
        endedAtIso: new Date(endTime).toISOString(),
        attributes: {
          'task.type': 'debug',
          'task.priority': 'low',
          'debug.enabled': true,
        },
      };
      
      await processWithProcessors(processors, spanInfo);
      addSpanEvent('debug.logged', { file: 'debug-spans.jsonl' });
      console.log('  Completed (processed by debug processor)');
    });

    // Analytics task - with console analytics
    console.log('\n3. Analytics Task:');
    await respan.withTask({ name: 'analytics_task' }, async () => {
      const startTime = Date.now();
      updateCurrentSpan({
        attributes: {
          'task.type': 'analytics',
          'task.priority': 'high',
          'analytics.enabled': true,
        },
      });
      addSpanEvent('analytics.started', { metrics: 'enabled' });
      
      await new Promise((resolve) => setTimeout(resolve, 80));
      
      const endTime = Date.now();
      const spanInfo = {
        name: 'analytics_task',
        type: 'analytics',
        durationMs: endTime - startTime,
        metrics: { processed: 42, errors: 0 },
        startedAtIso: new Date(startTime).toISOString(),
        endedAtIso: new Date(endTime).toISOString(),
        attributes: {
          'task.type': 'analytics',
          'task.priority': 'high',
          'analytics.enabled': true,
        },
      };
      
      await processWithProcessors(processors, spanInfo);
      addSpanEvent('analytics.completed', { records: 42 });
      console.log('  Completed (processed by analytics processor)');
    });

    // Slow task - demonstrates long-running operation
    console.log('\n4. Slow Task (long-running):');
    await respan.withTask({ name: 'slow_task' }, async () => {
      const startTime = Date.now();
      updateCurrentSpan({
        attributes: {
          'task.type': 'slow',
          'task.priority': 'low',
          'performance.warning': true,
        },
      });
      addSpanEvent('slow.task.started', { expected_duration: '200ms' });
      
      await new Promise((resolve) => setTimeout(resolve, 200));
      
      const endTime = Date.now();
      const duration = endTime - startTime;
      const spanInfo = {
        name: 'slow_task',
        type: 'slow',
        durationMs: duration,
        warning: duration > 100 ? 'Exceeded threshold' : null,
        startedAtIso: new Date(startTime).toISOString(),
        endedAtIso: new Date(endTime).toISOString(),
        attributes: {
          'task.type': 'slow',
          'task.priority': 'low',
          'performance.warning': true,
        },
      };
      
      await processWithProcessors(processors, spanInfo);
      addSpanEvent('slow.task.completed', { actual_duration: duration });
      console.log(`  Completed in ${duration}ms (processed by slow processor)`);
    });
  });

  console.log('\nShutting down...');
  await respan.shutdown();
  console.log('Span tracking demo completed.');
  console.log('\nCheck these files for logged spans:');
  console.log('   - ./debug-spans.jsonl');
  console.log('   - ./analytics-spans.jsonl');
  console.log('   - ./slow-spans.jsonl');
}

if (import.meta.url === `file://${process.argv[1]}`) {
  runMultiProcessorDemo().catch(console.error);
}

export { runMultiProcessorDemo, logSpanToFile, logSpanToConsole };

File: python/tracing/respan-tracing-sdk/multi_processor.py
#!/usr/bin/env python3
"""
Multi-Processor - KeywordsAI Tracing SDK
Demonstrates: routing spans to multiple destinations with add_processor()
"""
import os
import json
import asyncio
from pathlib import Path
from datetime import datetime
from typing import Sequence
from dotenv import load_dotenv

env_path = Path(__file__).parent / ".env"
load_dotenv(env_path, override=True)

from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult
from keywordsai_tracing import KeywordsAITelemetry, get_client
from keywordsai_tracing.decorators import workflow, task
from keywordsai_tracing.instruments import Instruments


class FileExporter(SpanExporter):
    """Writes spans to a JSONL file."""
    def __init__(self, filepath: str, name: str = "file"):
        self.filepath = filepath
        self.name = name
        Path(filepath).parent.mkdir(parents=True, exist_ok=True)
        open(filepath, "w").close()

    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
        with open(self.filepath, "a") as f:
            for span in spans:
                data = {"name": span.name, "exporter": self.name, "time": datetime.now().isoformat()}
                f.write(json.dumps(data) + "\n")
        return SpanExportResult.SUCCESS

    def shutdown(self) -> None:
        pass

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return True


keywords_ai = KeywordsAITelemetry(
    app_name="multi-processor",
    api_key=os.getenv("KEYWORDSAI_API_KEY"),
    is_batching_enabled=False,
    block_instruments={Instruments.REQUESTS, Instruments.URLLIB3},
)

# Add custom processors
keywords_ai.add_processor(exporter=FileExporter("./output/debug-spans.jsonl", "debug"), name="debug")
keywords_ai.add_processor(exporter=FileExporter("./output/analytics-spans.jsonl", "analytics"), name="analytics")


@task(name="normal_task")
async def normal_task():
    """Goes to default KeywordsAI processor."""
    await asyncio.sleep(0.05)
    return "normal"


@task(name="debug_task", processors="debug")
async def debug_task():
    """Goes to debug processor only."""
    client = get_client()
    if client:
        client.add_event("debug.checkpoint")
    await asyncio.sleep(0.05)
    return "debug"


@task(name="analytics_task", processors="analytics")
async def analytics_task():
    """Goes to analytics processor only."""
    await asyncio.sleep(0.05)
    return "analytics"


@task(name="multi_task", processors=["debug", "analytics"])
async def multi_task():
    """Goes to both debug and analytics processors."""
    await asyncio.sleep(0.05)
    return "multi"


@workflow(name="multi_processor_workflow")
async def run_demo():
    """Demonstrate processor routing."""
    results = {}
    results["normal"] = await normal_task()
    results["debug"] = await debug_task()
    results["analytics"] = await analytics_task()
    results["multi"] = await multi_task()
    return results


async def main():
    print("=" * 50)
    print("Multi-Processor Demo")
    print("=" * 50)
    print("Routing: default=KeywordsAI, debug=file, analytics=file\n")

    try:
        results = await run_demo()
        for k, v in results.items():
            print(f"  {k}: {v}")
    finally:
        keywords_ai.flush()
        await asyncio.sleep(1)

    print("\nOutput files:")
    print("  ./output/debug-spans.jsonl")
    print("  ./output/analytics-spans.jsonl")
    print("Done.")


if __name__ == "__main__":
    asyncio.run(main())
